protocol/l1-upgrades.md

```diff
@@ -0,0 +1,305 @@
+      // 1. Call the `SuperchainProxyAdmin` to update the implementation to the `Populator` contract and call `Populator.upgrade()`.
+      // 2. Update the implementation to the final implementation contract.
+  }
+  // Return ownership of the `SuperchainProxyAdmin` to the Upgrade Controller (`msg.sender`)
+      //    `SuperchainProxyAdmin`. The unique upgrade call encoding required for the non-standard proxies
+      //    should be handled here.
+}
+```
+
+**Note:**A new getter should be added to the `SystemConfig` to perform this in a single call.
+
+## L1ChugsplashProxy Considerations
+
+The L1 Chugsplash proxy is much more expensive to upgrade, as it is done by calling
+[`setCode()`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/legacy/L1ChugSplashProxy.sol#L92)
+which creates a new implementation contract. In this sense the L1StandardBridge is not truly MCP-L1
+ready.
+
+This will required an intermediate proxy to enable pointing to a single implementation contract,
+options for this intermediate proxy include:
+
+1. **Cheaper upgrades, costlier bridging:** a onetime upgrade which sets the implementation for all L1
+   Standard bridges to an intermediate standard Proxy contract. From then on the standard proxy
+   will have its implementation upgraded.
+   This is a tradeoff between token bridging costs and upgrade
+   costs. It may also add complexity in the form of needing to track a new contract for each OP
+   Chain.
+2. **Middle ground:** each time the implementation is changed, `setCode()` could deploy a minimal
+   intermediate immutable proxy which hardcodes the address of the implementation. This reduces the
+   size of the contract, is more gas efficient (no `sload` in the intermediate proxy) and avoids the
+   need to track the intermediate proxies.
+3. **Costlier upgrades, cheaper bridging:** deploy a new but identical implementation for every
+   L1 Standard Bridge. Off chain analysis can easily verify that the implementatin matches the
+   governance approved bytecode.
+
+Further gas cost analysis and product discussion will be required to make a final decision.
+
+## Development and Testing considerations
+
+This solution should be implemented such that it both enables and _requires_ developers to implement
+the upgrade path from the most recent release to the one curently being developed. This means it
+will be necessary to have a deployment of the previous release available on the current commit, in
+order to test the upgrade path.
+
+This can be achieved by storing an L1 Genesis representation of the latest release in the monorepo.
+This genesis file can be generated by from a state dump of the currently deployed test system. The
+test suite should then be modified so that each test runs twice, one against a system freshly
+deployed by the OPCM, and another which was upgraded from the previous release. In CI these two
+major test groups can be run in parallel.
```

Comment:
Having to run the full test suite twice is a big downside that we don't have for the alternate approach where we keep the `initialize` method. Something to discuss in the "Alternatives considered" section

Reply:
Instead of duplicating the tests, I think it should be sufficient to prove that the two paths result in contracts with identical config.

Reply:
Hmm intuitively we want to test both fresh deployments & upgrades so I am not super duper concerned about this, especially because they can run in parallel

Reply:
> Hmm intuitively we want to test both fresh deployments & upgrades so I am not super duper concerned about this, especially because they can run in parallel

OK, I've included this in the document: 

```
All tests should be executed against both a freshly deployed and upgraded system.

**Alternatives considered:**

- Rather than duplicated tests, new tests could ensure that given the same configuration as inputs,
  a fresh chain has the same resulting configuration as an upgraded chain. Duplicating tests was
  deemed acceptable as they are parallelizable.
```